File: packages/python-sdk/wusul/http_client.py
"""HTTP client for making authenticated requests to the Wusul API."""
from typing import Any, Dict, Optional
from urllib.parse import urlencode

import requests

from .auth import create_auth_headers, create_get_auth_headers


class HttpClient:
    """HTTP client for making authenticated requests to the Wusul API."""

    def __init__(self, account_id: str, shared_secret: str, base_url: str, timeout: int):
        """
        Initialize the HTTP client.

        Args:
            account_id: Your Wusul account ID
            shared_secret: Your Wusul shared secret
            base_url: Base URL for the API
            timeout: Request timeout in seconds
        """
        self.account_id = account_id
        self.shared_secret = shared_secret
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout / 1000.0  # Convert milliseconds to seconds

        self.session = requests.Session()
        self.session.headers.update({"Content-Type": "application/json"})

    def _handle_response(self, response: requests.Response) -> Any:
        """
        Handle API response and extract data.

        Args:
            response: The response object

        Returns:
            Response data

        Raises:
            Exception: If the response indicates an error
        """
        try:
            response.raise_for_status()
        except requests.exceptions.HTTPError as e:
            error_message = None
            try:
                error_data = response.json()
                error_message = error_data.get("error") or error_data.get("message")
            except Exception:
                pass

            if error_message:
                raise Exception(f"Wusul API Error ({response.status_code}): {error_message}")
            else:
                raise Exception(f"Wusul API Error ({response.status_code}): {str(e)}")

        try:
            data = response.json()
            return data.get("data", data)
        except Exception:
            return response.text

    def get(self, url: str, sig_payload: Optional[Dict[str, Any]] = None) -> Any:
        """
        Make a GET request.

        Args:
            url: The endpoint URL
            sig_payload: Optional signature payload

        Returns:
            Response data
        """
        headers, encoded_payload = create_get_auth_headers(
            self.account_id, self.shared_secret, sig_payload
        )

        full_url = f"{self.base_url}{url}"
        params = {"sig_payload": encoded_payload}

        try:
            response = self.session.get(full_url, headers=headers, params=params, timeout=self.timeout)
            return self._handle_response(response)
        except requests.exceptions.Timeout:
            raise Exception("Request timeout - no response received from Wusul API")
        except requests.exceptions.RequestException as e:
            raise Exception(f"Request error: {str(e)}")

    def post(self, url: str, data: Optional[Any] = None) -> Any:
        """
        Make a POST request.

        Args:
            url: The endpoint URL
            data: Optional request body

        Returns:
            Response data
        """
        headers = create_auth_headers(self.account_id, self.shared_secret, data)
        full_url = f"{self.base_url}{url}"

        try:
            response = self.session.post(
                full_url, json=data, headers=headers, timeout=self.timeout
            )
            return self._handle_response(response)
        except requests.exceptions.Timeout:
            raise Exception("Request timeout - no response received from Wusul API")
        except requests.exceptions.RequestException as e:
            raise Exception(f"Request error: {str(e)}")

    def patch(self, url: str, data: Optional[Any] = None) -> Any:
        """
        Make a PATCH request.

        Args:
            url: The endpoint URL
            data: Optional request body

        Returns:
            Response data
        """
        headers = create_auth_headers(self.account_id, self.shared_secret, data)
        full_url = f"{self.base_url}{url}"

        try:
            response = self.session.patch(
                full_url, json=data, headers=headers, timeout=self.timeout
            )
            return self._handle_response(response)
        except requests.exceptions.Timeout:
            raise Exception("Request timeout - no response received from Wusul API")
        except requests.exceptions.RequestException as e:
            raise Exception(f"Request error: {str(e)}")

    def delete(self, url: str) -> Any:
        """
        Make a DELETE request.

        Args:
            url: The endpoint URL

        Returns:
            Response data
        """
        headers = create_auth_headers(self.account_id, self.shared_secret)
        full_url = f"{self.base_url}{url}"

        try:
            response = self.session.delete(full_url, headers=headers, timeout=self.timeout)
            return self._handle_response(response)
        except requests.exceptions.Timeout:
            raise Exception("Request timeout - no response received from Wusul API")
        except requests.exceptions.RequestException as e:
            raise Exception(f"Request error: {str(e)}")

File: packages/php-sdk/src/HttpClient.php
<?php

namespace Wusul;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\Exception\RequestException;

/**
 * HTTP client for making authenticated requests to the Wusul API
 */
class HttpClient
{
    private GuzzleClient $client;
    private string $accountId;
    private string $sharedSecret;

    /**
     * Creates a new HTTP client instance
     *
     * @param string $accountId The account ID
     * @param string $sharedSecret The shared secret
     * @param string $baseUrl The base URL for the API
     * @param int $timeout Request timeout in seconds
     */
    public function __construct(string $accountId, string $sharedSecret, string $baseUrl, int $timeout)
    {
        $this->accountId = $accountId;
        $this->sharedSecret = $sharedSecret;

        $this->client = new GuzzleClient([
            'base_uri' => $baseUrl,
            'timeout' => $timeout,
            'headers' => [
                'Content-Type' => 'application/json',
            ],
        ]);
    }

    /**
     * Makes a GET request
     *
     * @param string $url The endpoint URL
     * @param array|null $sigPayload Optional signature payload
     * @return mixed The response data
     * @throws \Exception
     */
    public function get(string $url, ?array $sigPayload = null)
    {
        try {
            $authData = Auth::createGetAuthHeaders($this->accountId, $this->sharedSecret, $sigPayload);

            $response = $this->client->get($url, [
                'headers' => $authData['headers'],
                'query' => [
                    'sig_payload' => $authData['sigPayload'],
                ],
            ]);

            $data = json_decode($response->getBody()->getContents(), true);
            return $data['data'] ?? $data;
        } catch (RequestException $e) {
            $this->handleException($e);
        } catch (GuzzleException $e) {
            throw new \Exception('Request error: ' . $e->getMessage());
        }
    }

    /**
     * Makes a POST request
     *
     * @param string $url The endpoint URL
     * @param array|null $data The request body data
     * @return mixed The response data
     * @throws \Exception
     */
    public function post(string $url, ?array $data = null)
    {
        try {
            $headers = Auth::createAuthHeaders($this->accountId, $this->sharedSecret, $data);

            $response = $this->client->post($url, [
                'headers' => $headers,
                'json' => $data,
            ]);

            $responseData = json_decode($response->getBody()->getContents(), true);
            return $responseData['data'] ?? $responseData;
        } catch (RequestException $e) {
            $this->handleException($e);
        } catch (GuzzleException $e) {
            throw new \Exception('Request error: ' . $e->getMessage());
        }
    }

    /**
     * Makes a PATCH request
     *
     * @param string $url The endpoint URL
     * @param array|null $data The request body data
     * @return mixed The response data
     * @throws \Exception
     */
    public function patch(string $url, ?array $data = null)
    {
        try {
            $headers = Auth::createAuthHeaders($this->accountId, $this->sharedSecret, $data);

            $response = $this->client->patch($url, [
                'headers' => $headers,
                'json' => $data,
            ]);

            $responseData = json_decode($response->getBody()->getContents(), true);
            return $responseData['data'] ?? $responseData;
        } catch (RequestException $e) {
            $this->handleException($e);
        } catch (GuzzleException $e) {
            throw new \Exception('Request error: ' . $e->getMessage());
        }
    }

    /**
     * Makes a DELETE request
     *
     * @param string $url The endpoint URL
     * @return mixed The response data
     * @throws \Exception
     */
    public function delete(string $url)
    {
        try {
            $headers = Auth::createAuthHeaders($this->accountId, $this->sharedSecret);

            $response = $this->client->delete($url, [
                'headers' => $headers,
            ]);

            $responseData = json_decode($response->getBody()->getContents(), true);
            return $responseData['data'] ?? $responseData;
        } catch (RequestException $e) {
            $this->handleException($e);
        } catch (GuzzleException $e) {
            throw new \Exception('Request error: ' . $e->getMessage());
        }
    }

    /**
     * Handles exceptions from HTTP requests
     *
     * @param RequestException $e The exception
     * @throws \Exception
     */
    private function handleException(RequestException $e): void
    {
        if ($e->hasResponse()) {
            $response = $e->getResponse();
            $body = json_decode($response->getBody()->getContents(), true);
            $errorMessage = $body['error'] ?? $body['message'] ?? $e->getMessage();
            throw new \Exception('Wusul API Error (' . $response->getStatusCode() . '): ' . $errorMessage);
        }

        throw new \Exception('No response received from Wusul API');
    }
}
